Adding documentation to solve the kubefed-admission-webhook timeout

queeno opened the pull request:
In GKE, the kubefed-controller fails to start up because it cannot create
new kubefed objects. This is because the API server attempts to
communicate to the Pod directly rather than going through the service.
This is intended behaviour when the API server runs with the
`--enable-aggregator-routing=true`.

An additional firewall rule within the VPC network firewall rules must
be added. The GKE installation guide has been extended to inform the
users of this issue.

Resolves #1024

queeno commented:
CLA signed

queeno commented:
Invalid commit message fixed

queeno commented:
/meow
/remove-invalid-commit-message

queeno commented:
/retest

queeno commented:
/remove-label do-not-merge/invalid-commit-message

queeno commented:
/assign @shashidharatd

queeno commented:
/pony

font commented on docs/environments/gke.md:
```diff
@@ -63,3 +62,74 @@ kubectl create clusterrolebinding myname-cluster-admin-binding --clusterrole=clu
+
+### Background
+
+The GCP Kubernetes control plane is deployed in a Google-managed account.
```
s/GCP/GKE/

font commented on docs/environments/gke.md:
```diff
@@ -63,3 +62,74 @@ kubectl create clusterrolebinding myname-cluster-admin-binding --clusterrole=clu
+1. A network peering connection to allow the master to communicate to the node pools.
+1. A firewall rule to allow ports 443 and 10250 from the master to the node pools.
+
+The GKE API Server operates with [`--enable-aggregator-routing=true`](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/gci/configure-helper.sh#L1586). This means that the Service IPs are internally translated to endpoint IPs and routed to the relevant pods.
```
Is there an official GKE doc that describes this flag being set by default which could be linked instead? If not, the link provided here to a specific line will undoubtedly change as it's not a permalink.

queeno replied:
Very good point. Unfortunately there isn't - as far as I know. However, I would link the [kubernetes issue](https://github.com/kubernetes/kubernetes/issues/79739) I opened with them. They promised to create some documentation to inform GCP users regarding this behaviour.

font commented on docs/environments/gke.md:
```diff
@@ -63,3 +62,74 @@ kubectl create clusterrolebinding myname-cluster-admin-binding --clusterrole=clu
+
+The GKE API Server operates with [`--enable-aggregator-routing=true`](https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/gci/configure-helper.sh#L1586). This means that the Service IPs are internally translated to endpoint IPs and routed to the relevant pods.
+
+A number of [webhooks are deployed](https://github.com/kubernetes-sigs/kubefed/blob/master/charts/kubefed/charts/controllermanager/templates/webhook.yaml) during the kubfed helm installation, alongside the `kubefed-admission-webhook` application. Every time a new kubefed object is created, the webhook causes the API server to query the `kubefed-admission-webhook` service.
```
Please use relative links.

font replied:
Additional specificity:

> A number of webhook configurations

font replied:
Replace

> Every time a new kubefed object is created

with:

> Every time a validated kubefed object is created or updated

font commented on docs/environments/gke.md:
```diff
@@ -63,3 +62,74 @@ kubectl create clusterrolebinding myname-cluster-admin-binding --clusterrole=clu
+
+The GCP Kubernetes control plane is deployed in a Google-managed account.
+
+During cluster provisioning, GKE automatically creates:
```
Are there any links to the GKE docs describing this? If so, please add references.

font commented on docs/environments/gke.md:
```diff
@@ -63,3 +62,74 @@ kubectl create clusterrolebinding myname-cluster-admin-binding --clusterrole=clu
+
+A number of [webhooks are deployed](https://github.com/kubernetes-sigs/kubefed/blob/master/charts/kubefed/charts/controllermanager/templates/webhook.yaml) during the kubfed helm installation, alongside the `kubefed-admission-webhook` application. Every time a new kubefed object is created, the webhook causes the API server to query the `kubefed-admission-webhook` service.
+
+Because of the `enable-aggregator-routing` being turned on, the request is directed from the API server to the pod running the `kubefed-admission-webhook` application directly, on port 8443. This traffic needs to be allowed through the GCP firewall in order for the webhook to succeed.
```
s/webhook to succeed/webhook query to succeed/

font commented on docs/environments/gke.md:
```diff
@@ -63,3 +62,74 @@ kubectl create clusterrolebinding myname-cluster-admin-binding --clusterrole=clu
+
+### Issue
+
+The kubefed-controller pod fails to start up when attempting to create a KubeFedConfig in kubernetes. A timeout error is observed.
```
s/kubefed-controller pod/`kubefed-controller-manager`/

font replied:
s/KubeFedConfig in kubernetes/`KubeFedConfig` resource/

font commented on docs/environments/gke.md:
````diff
@@ -63,3 +62,74 @@ kubectl create clusterrolebinding myname-cluster-admin-binding --clusterrole=clu
+
+```
+Error updating KubeFedConfig "kube-federation-system/kubefed": Timeout: request did not complete within requested timeout 30s
+```
````
Consider also linking to the issue #1024 you filed.

font commented on docs/environments/gke.md:
````diff
@@ -63,3 +62,74 @@ kubectl create clusterrolebinding myname-cluster-admin-binding --clusterrole=clu
+#### GCloud
+
+```bash
+gcloud compute firewall-rules create "allow-apiserver-to-kubefed-admission-webhook-8443" --allow tcp:8443 \
````
This firewall isn't specific to `kubefed` so consider naming it generically e.g. `allow-apiserver-to-admission-webhooks-8443`

queeno replied:
Good point :+1:

font commented on docs/environments/gke.md:
````diff
@@ -63,3 +62,74 @@ kubectl create clusterrolebinding myname-cluster-admin-binding --clusterrole=clu
+```bash
+gcloud compute firewall-rules create "allow-apiserver-to-kubefed-admission-webhook-8443" --allow tcp:8443 \
+      --network="[VPC_NETWORK]" \
+      --source-ranges="[MASTER_IPV4_CIDR_BLOCK]" \
````
To go one step further, consider adding steps for the user to run CLI commands to set environment variables for these different values. For example:

```bash
VPC_NETWORK=$(gcloud ...)
MASTER_IPV4_CIDR_BLOCK=$(gcloud ...)
...
```
This would allow the user to copy/paste the commands as much as possible.

font commented on docs/environments/gke.md:
````diff
@@ -63,3 +62,74 @@ kubectl create clusterrolebinding myname-cluster-admin-binding --clusterrole=clu
+      --direction INGRESS
+```
+
+#### Terraform
````
I don't think we need to provide and support a `Terraform` solution for this. The `gcloud` solution should suffice.

queeno replied:
👍

font commented on docs/environments/gke.md:
```diff
@@ -63,3 +62,74 @@ kubectl create clusterrolebinding myname-cluster-admin-binding --clusterrole=clu
+
+Because of the `enable-aggregator-routing` being turned on, the request is directed from the API server to the pod running the `kubefed-admission-webhook` application directly, on port 8443. This traffic needs to be allowed through the GCP firewall in order for the webhook to succeed.
+
+### Issue
```
Consider moving `Issue` before `Background`.

queeno commented:
Hi @font 
Thanks very much for commenting in this PR I have accepted all your observations.
I have also found out that this issue only occurs in private GKE clusters - as when the node pools are public the master plane talks to the node pool via an [SSH tunnel](https://github.com/kubernetes/kubernetes/issues/79739#issuecomment-509813068).
I have expanded the documentation to mention this.
Should you have further remarks, please let me know.

font commented on docs/environments/gke.md:
```diff
@@ -63,3 +63,69 @@ kubectl create clusterrolebinding myname-cluster-admin-binding --clusterrole=clu
+This means that the Service IPs are internally translated to endpoint IPs and
+routed to the relevant pods.
+
+A number of [webhook configurations are deployed](charts/controllermanager/templates/webhook.yaml)
```
This link needs an additional `../../` prefix.

font commented on docs/environments/gke.md:
```diff
@@ -63,3 +63,69 @@ kubectl create clusterrolebinding myname-cluster-admin-binding --clusterrole=clu
+      --network="$VPC_NETWORK" \
+      --source-ranges="$MASTER_IPV4_CIDR_BLOCK" \
+      --target-tags="$NODE_POOLS_TARGET_TAGS" \
+      --description="Allow apiserver access to kubefed-admission-webhook pod on port 8443" \
```
Since we made the name generic, let's make this description generic too.

queeno commented:
Hi @font 
Apologies for the short delay. Thanks for your comments.
I have addressed them all, committed and rebased against current master.
Hope the PR is good to be merged.
Thanks!

font commented on docs/environments/gke.md:
```diff
@@ -63,3 +63,69 @@ kubectl create clusterrolebinding myname-cluster-admin-binding --clusterrole=clu
+This means that the Service IPs are internally translated to endpoint IPs and
+routed to the relevant pods.
+
+A number of [webhook configurations are deployed](../../charts/controllermanager/templates/webhook.yaml)
```
Please verify the link works as it is missing an additional `charts/kubefed`.

queeno replied:
Sorry, I didn't test this after changing it. It should work now.

queeno commented:
And squashed!
